packages/dds/tree/src/simple-tree/tree.ts

```diff
@@ -135,29 +166,74 @@ export interface TreeView<TSchema extends ImplicitFieldSchema> extends IDisposab
 	 */
 	upgradeSchema(): void;
 
+	/**
+	 * Initialize the tree, setting the stored schema to match this view's schema and setting the tree content.
+	 *
+	 * Only valid to call when this view's {@link SchemaCompatibilityStatus.canInitialize} is true.
+	 *
+	 * Applications should typically call this function before attaching a `SharedTree`.
+	 * @param content - The content to initialize the tree with.
+	 */
+	initialize(content: InsertableTreeFieldFromImplicitField<TSchema>): void;
+
 	/**
 	 * Events for the tree.
 	 */
 	readonly events: ISubscribable<TreeViewEvents>;
 }
 
 /**
- * Information about how a view schema was incompatible.
+ * Information about a view schema's compatibility with the document's stored schema.
  * @public
  */
-export interface SchemaIncompatible {
+export interface SchemaCompatibilityStatus {
 	/**
-	 * True iff the view schema supports all possible documents permitted by the stored schema.
+	 * Whether the view schema is an exact match to the stored schema.
+	 */
+	readonly isExactMatch: boolean;
+
+	/**
+	 * Whether the current view schema is sufficiently compatible with the stored schema to allow viewing tree data.
+	 * This is true when the documents allowed by the view schema are a subset of those allowed by the stored schema.
+	 * If false, {@link TreeView.root} will throw upon access.
+	 *
+	 * Be aware that even when this is true, application logic may not correctly tolerate the documents allowable by the stored schema!
+	 * For example, if the stored schema allows types that the view schema does not, and the application doesn't have fallback logic
+	 * for unrecognized types as part of a field, it may throw when trying to read the document!
```

Comment:
This doesn't align with my assumptions for what "canView" would mean.

I was expecting "canView" to mean that the set of possible documents supported by the viewSchema is a superset of the documents supported by the stored schema. Or equivalently that the document content is known to be compatible with the view schema.

This comment seems to imply this is not the case (my interpretation of it implies that canView may be true when the document may have data which is not compatible with the view schema). Are you changing the cases in which its allowed to view the document contents? If so, I'll need to think about this more deeply to understand the implications.

Reply:
Yes. This was one of the main bits of feedback from the API review. I expounded on this topic a bit in the evolution readme, but when compared to the previous design, it comes down to a desire to not be overly prescriptive on how applications choose to maintain compatibility within their ecosystem.

Notably, forcing exact schema match and consensus across clients means ecosystems must coordinate & ship multiple schemas whenever they want to make a schema change, assuming they want to preserve collaboration compatibility. This could come as a perceived regression from existing Fluid customers, who may be used to making backward-compatible changes (such as adding an optional field that has truly optional semantics--old client can collab with new client without issue and just ignores the new data) when using other DDSes like SharedMap/SharedDirectory.

By allowing view schema which supports a subset of documents that the stored schema supports, we ensure that any edits such a client can make are valid, so they're unable to corrupt the document. Since SharedTree internals generally operate using the stored schema, they end up working ok.

Collaborating without an exact schema match runs the risk that application code chokes at runtime due to unexpected data it cannot handle. Future plans involve letting the application specify metadata with their schema as well as some APIs for assessing the type of incompatibility, and applications typically will want to use those to have robust mechanisms in this area (if they want to allow collaboration with different schemas). The readme filed I added here is meant to outline some example policies (though the APIs they use aren't currently implemented, which is noted).

If an application wants to opt for the previous--stricter--behavior, they can always refuse to open unless the compatibility indicates an exact match.

Reply:
> By allowing view schema which supports a subset of documents that the stored schema supports, we ensure that any edits such a client can make are valid, so they're unable to corrupt the document. Since SharedTree internals generally operate using the stored schema, they end up working ok.

That sounds like its safe to edit, but might not be able to view. Calling that "can view" seems odd to me.

We can detect cases where edits are known to be safe, but reading is not, and cases where reading is known to be safe but editing might not be (and cases where neither is compatible).

In my (old) design the assumption was that supporting stored schema more general than your view schema would be done by providing adapters. I think your proposed approach here is equivalent to what I was imagining except that you implicitly get adapters which subset node content when valid, and maybe skip items in sequences (and optional fields?) that arn't supported.


This is more complicated that it might appear, since what happens if someone moves a range in a sequence where some of the items are hidden to them, or enumerates an object node to copy it when some of its fields don't exist in their view? The simple-tree and flex-tree APIs currently assume/require that the forest content is in schema with the view schema, and having extra unexpected content would violate that assumption. With care and a few targeted changes (and some docs updates) I suspect we could change that assumption to accommodate some of those cases but I don't think this PR includes those changes.

While I agree there is a problem we need to solve here, and I think you are pretty close to a good initial solution that actually allows cross version collaboration, I don't think you have addressed all the issues that lead to me getting support for actual cross version colab with different view schema getting punted and focusing on the whole monotonic matching schema simplified setup we have currently.

Reply:
FYI--I plan on reducing the scope of "canView" in the 2.0 release to only encompass view schemas which have extra additional properties over the stored schema. This should cover the concern around 'strict updates' being a perceived regression from people familiar with preserving backward compatibility for things like `SharedMap/SharedDirectory` (e.g. new versions can add optional data to new keys, but if they change the expected schema for an existing key to a different object, they'll generally expect crashes), while still leaving the door open for us to allow support for things like out-of-schema adapters for forward compatibility in fields that applications anticipate content types to be added to over time.

Current plan is to accomplish this by moving `ViewSchema.checkCompatibility` to use helpers which report more information about discrepancies between the two schemas rather than just `allowRepoSuperset`.

fwiw, I don't think this plan will necessitate many code changes to simple-tree layer based on initial testing: seems like object reflection APIs already only report properties within the view schema, even when there is additional optional data, which is reasonable enough. It does mean this case you brought up can give potentially odd results from an app perspective:

> enumerates an object node to copy it when some of its fields don't exist in their view

but I think that limitation is fine, and I'll call it out in the app-facing documentation on compatibility considerations. You can hit similar stuff even if you don't do object enumeration (e.g. if you construct a new object for insertion using the properties you know the old object had, you'll miss extra optional data you're unaware of).

In terms of getting toward that end state, I'm going to update this PR so that it doesn't change underlying semantics of which schema we allow the application to view the document with, but still has the bulk of the other API changes. We can then add support for opening documents using a view schema which omits some optional fields present in the stored schema in a separate PR. This should unblock other efforts like migrating our tests to use `viewWith` rather than `schematize` and removing `schematize`.